sql_extensions/migrations/special-queries.sql

```diff
@@ -47,15 +47,54 @@ WHERE
 ORDER BY
     (data -> 'market_metadata' ->> 'market_id')::NUMERIC,
     (data ->> 'transaction_version')::NUMERIC DESC,
-    (data ->> 'event_index')::NUMERIC DESC;
+    (data -> 'state_metadata' ->> 'market_nonce')::NUMERIC DESC;
 
 CREATE OR REPLACE FUNCTION UPDATE_LATEST_STATE()
 RETURNS TRIGGER AS $$
 BEGIN
-  DELETE FROM inbox_latest_state
-  WHERE (inbox_latest_state.data->'market_metadata'->>'market_id')::numeric = (NEW.data->'market_metadata'->>'market_id')::numeric
-  AND (inbox_latest_state.data->'state_metadata'->>'bump_time')::numeric <= (NEW.data->'state_metadata'->>'bump_time')::numeric;
-  INSERT INTO inbox_latest_state SELECT NEW.*;
+  IF EXISTS (
+    SELECT 1
+    FROM inbox_latest_state
+    WHERE (inbox_latest_state.data->'market_metadata'->>'market_id')::numeric = (NEW.data->'market_metadata'->>'market_id')::numeric
+      AND (inbox_latest_state.data->'state_metadata'->>'market_nonce')::numeric < (NEW.data->'state_metadata'->>'market_nonce')::numeric
+  ) THEN
+    RETURN NEW;
+  END IF;
+
+  INSERT INTO inbox_latest_state (
+    sequence_number,
+    creation_number,
+    account_address,
+    transaction_version,
+    transaction_block_height,
+    type,
+    data,
+    inserted_at,
+    event_index,
+    indexed_type,
+    market_id
+  )
+  VALUES (
+    NEW.sequence_number,
+    NEW.creation_number,
+    NEW.account_address,
+    NEW.transaction_version,
+    NEW.transaction_block_height,
+    NEW.type,
+    NEW.data,
+    NEW.inserted_at,
+    NEW.event_index,
+    NEW.indexed_type,
+    (NEW.data -> 'market_metadata' ->> 'market_id')::numeric
+  )
+  ON CONFLICT (market_id) DO UPDATE SET
+    transaction_version = EXCLUDED.transaction_version,
+    transaction_block_height = EXCLUDED.transaction_block_height,
+    data = EXCLUDED.data,
+    inserted_at = EXCLUDED.inserted_at,
+    event_index = EXCLUDED.event_index
+  WHERE (inbox_latest_state.data->'state_metadata'->>'market_nonce')::numeric < (EXCLUDED.data->'state_metadata'->>'market_nonce')::numeric;
```

Comment:
Oh, or is the `IF` what fixed the deadlock problem ? If that is the case, just remove the `WHERE` here.

Reply:
Actually not sure about this. If the `IF` and `INSERT` statements do not happen in an atomic way, the `WHERE` should be kept, but then I think the `IF` doesn't make sense anymore since it isn't preventing a deadlock. I would definitely try and remove that if and see if it still works.

Reply:
After some testing, it seems that this is subject to race conditions. I will have to find a way to avoid this.